source/particle/property/cpo.cc

```diff
@@ -0,0 +1,1124 @@
+                                             std::vector<std::vector<Tensor<2,3> > > &a_cosine_matrices_grains_derivatives) const
+      {
+        // See class header information for layout information
+        store_particle_data(cpo_data_position,
+                            data,
+                            deformation_type,
+                            volume_fraction_mineral,
+                            volume_fractions_grains,
+                            a_cosine_matrices_grains);
+
+        // now store the derivatives if needed. They are added after all the other data.
+        if (this->advection_method == AdvectionMethod::CrankNicolson)
+          {
+            for (size_t mineral_i = 0; mineral_i < n_minerals; mineral_i++)
+              {
+                Assert(volume_fractions_grains_derivatives.size() == n_minerals, ExcMessage("Internal error: volume_fractions_olivine_derivatives is not the same as n_minerals."));
+                Assert(a_cosine_matrices_grains_derivatives.size() == n_minerals, ExcMessage("Internal error: a_cosine_matrices_olivine_derivatives is not the same as n_minerals."));
+
+                for (size_t mineral_i = 0; mineral_i < n_minerals; mineral_i++)
+                  {
+                    Assert(volume_fractions_grains_derivatives[mineral_i].size() == n_grains, ExcMessage("Internal error: volume_fractions_olivine_derivatives is not the same as n_grains."));
+                    Assert(a_cosine_matrices_grains_derivatives[mineral_i].size() == n_grains, ExcMessage("Internal error: a_cosine_matrices_olivine_derivatives is not the same as n_grains."));
+
+                    for (unsigned int grain_i = 0; grain_i < n_grains; ++grain_i)
+                      {
+                        // store volume fraction for olivine grains
+                        data[cpo_data_position + n_minerals * (n_grains * 10 + 2) + mineral_i * (n_grains * 10)  + grain_i * 10] = volume_fractions_grains_derivatives[mineral_i][grain_i];
+
+                        // store a_{ij} for olivine grains
+                        for (unsigned int iii = 0; iii < Tensor<2,3>::n_independent_components ; ++iii)
+                          {
+                            const dealii::TableIndices<2> index = Tensor<2,3>::unrolled_to_component_indices(iii);
+                            data[cpo_data_position + n_minerals * (n_grains * 10 + 2) + mineral_i * (n_grains * 10)  + grain_i * 10  + 1 + iii] = a_cosine_matrices_grains_derivatives[mineral_i][grain_i][index];
+                          }
+                      }
+                  }
+              }
+          }
+      }
+
+      template <int dim>
+      void
+      CPO<dim>::initialize_one_particle_property(const Point<dim> &,
+                                                 std::vector<double> &data) const
+      {
+        // the layout of the data vector per perticle is the following (note that for this plugin the following dim's are always 3):
```

Comment:
the 2d case just fills in zero's for the 3rth dimension of the strain-rate and velocity gradient tensors. To be honest, I havn't spend that much thought on 2D, and visualization of the results is also not trivial, because currently the aspect paraview output doesn't understand a 3D vector in a 2D model. I would be fine with just asserting the 2D case, since I see little value in 2D cpo, but I don't know how others think about that.